File: tools/auction-arbitrage/arb/calibration.py
"""The feedback loop: grade past predictions against what actually happened, then bend future
valuations toward reality.

Two tiers of evidence, weakest first:

1. HAMMER PRICES (free, automatic, every closed lot -- even ones you never bid on). HiBid publishes
   ``priceRealized``. A hammer price is not a resale value, so it cannot tell us we were too LOW.
   But it can prove we were too HIGH: if a lot's landed cost at the hammer meets or beats the net
   resale we predicted, then either the winner overpaid or -- far more often across many lots --
   our number was inflated. The share of lots where that happens is the ``overshoot_rate``, and it
   is a provably-wrong rate, not a guess.

2. REAL SALES (ground truth, entered when you actually sell something). actual sale / predicted mid.
   This overrides tier 1 as soon as there is enough of it.

Bias is clamped and needs a minimum sample, so a thin or unlucky week cannot swing the model.
"""
from __future__ import annotations

import statistics
import time
from typing import Any, Iterable

from .config import Settings

GLOBAL = "__all__"


def _median(xs: list[float]) -> float | None:
    return statistics.median(xs) if xs else None


def hammer_ratio(o: dict[str, Any]) -> float | None:
    """Landed cost at the hammer over the net resale we predicted. >= 1 means the deal was never there."""
    lah, net = o.get("landed_at_hammer"), o.get("predicted_net")
    if not lah or not net or net <= 0:
        return None
    return lah / net


def sale_ratio(o: dict[str, Any]) -> float | None:
    """Actual sale price over the mid we predicted. 1.0 is a perfect call."""
    sale, mid = o.get("sale_price"), o.get("predicted_mid")
    if not sale or not mid or mid <= 0:
        return None
    return sale / mid


def _calibrate_group(category: str, rows: list[dict[str, Any]], s: Settings, now: float) -> dict[str, Any]:
    hammers = [r for r in (hammer_ratio(o) for o in rows) if r is not None]
    sales = [r for r in (sale_ratio(o) for o in rows) if r is not None]
    med_hammer, med_sale = _median(hammers), _median(sales)
    overshoot = (sum(1 for r in hammers if r >= 1) / len(hammers)) if hammers else None
    mape = _median([abs(r - 1) for r in sales]) if sales else None

    bias, conf_factor, basis = 1.0, 1.0, "none"
    if len(sales) >= s.calibration_min_sales and med_sale is not None:
        basis = "sales"  # ground truth wins outright
        bias = med_sale
        if mape is not None:
            conf_factor = max(0.4, min(1.0, 1 - max(0.0, mape - 0.2)))
    elif len(hammers) >= s.calibration_min_closed and overshoot is not None:
        # No sales yet: we can only detect inflation, so only ever haircut on this basis.
        basis = "hammer"
        excess = max(0.0, overshoot - 0.25)  # a quarter of picks getting outbid is normal
        bias = min(1.0, 1 - excess)
        conf_factor = max(0.5, min(1.0, 1 - excess))

    bias = max(s.calibration_min_bias, min(s.calibration_max_bias, bias))
    return {
        "category": category, "n_closed": len(rows), "n_sold": len(sales),
        "median_hammer_ratio": med_hammer, "overshoot_rate": overshoot,
        "median_sale_ratio": med_sale, "sale_mape": mape,
        "bias": round(bias, 3), "confidence_factor": round(conf_factor, 3),
        "basis": basis, "updated_at": now,
    }


# ---------------------------------------------------------------------------- liquidity feedback
#
# The price loop above asks "was the number right". This one asks "did it ever actually sell", which is
# the question a $40 item sitting unviewed for three months answers differently.
#
# OBSERVED DAYS -- for the things that sold, listed_at to sale_at. Median observed over median predicted
# gives a per-category multiplier that stretches (or compresses) every future speed estimate.
#
# 30-DAY SELL RATE -- sold within 30 days, over everything that had a fair shot at 30 days. The
# denominator deliberately includes listings that are STILL SITTING, because those are the whole point:
# a loop that only learns from things that sold would conclude everything sells. A listing that has been
# up for 12 days and has not sold is not yet evidence either way, so it is excluded until day 30
# (right-censoring); one that has been up for 90 days unsold counts fully against the rate.


def observed_days(o: dict[str, Any]) -> float | None:
    """Days from listing to sale, for the ones that sold."""
    listed, sold = o.get("listed_at"), o.get("sale_at")
    if not listed or not sold:
        return None
    d = (sold - listed) / 86400
    return d if d >= 0 else None


def days_on_market(o: dict[str, Any], now: float) -> float | None:
    """Days a listing has been up, sold or not."""
    listed = o.get("listed_at")
    if not listed:
        return None
    d = ((o.get("sale_at") or now) - listed) / 86400
    return d if d >= 0 else None


def _calibrate_liquidity_group(category: str, rows: list[dict[str, Any]], s: Settings, now: float) -> dict[str, Any]:
    listed = [o for o in rows if o.get("listed_at")]
    sold = [o for o in listed if o.get("sale_at") and o.get("sale_price") is not None]
    obs = [d for d in (observed_days(o) for o in sold) if d is not None]
    # Compare like with like: only the items we both predicted and observed.
    pairs = [(observed_days(o), o["predicted_days"]) for o in sold
             if observed_days(o) is not None and o.get("predicted_days")]
    pairs = [(a, b) for a, b in pairs if b and b > 0]

    multiplier, basis = 1.0, "none"
    if len(pairs) >= s.liquidity_min_sales:
        m = _median([a / b for a, b in pairs])
        if m:
            multiplier, basis = max(0.3, min(4.0, m)), "measured"

    # Censoring: a listing only enters the 30-day denominator once it has had 30 days, or once it sold.
    eligible = [o for o in listed
                if (days_on_market(o, now) or 0) >= 30 or o.get("sale_at")]
    within30 = [o for o in eligible if (observed_days(o) or 1e9) <= 30]
    stuck = sum(1 for o in listed if not o.get("sale_at") and (days_on_market(o, now) or 0) >= 60)
    med_obs = _median(obs)
    med_pred = _median([b for _, b in pairs])
    return {
        "category": category, "n_listed": len(listed), "n_sold": len(sold),
        "observed_days": round(med_obs, 1) if med_obs is not None else None,
        "predicted_days": round(med_pred, 1) if med_pred is not None else None,
        "days_multiplier": round(multiplier, 3),
        "sell_rate_30d": round(len(within30) / len(eligible), 3) if eligible else None,
        "stuck": stuck, "basis": basis, "updated_at": now,
    }


def build_liquidity_report(outcomes: Iterable[dict[str, Any]], s: Settings, now: float) -> dict[str, Any]:
    rows = list(outcomes)
    by_cat: dict[str, list[dict[str, Any]]] = {}
    for o in rows:
        by_cat.setdefault(o.get("category") or "Uncategorized", []).append(o)
    categories = [g for g in (_calibrate_liquidity_group(c, r, s, now) for c, r in by_cat.items())
                  if g["n_listed"] > 0]
    categories.sort(key=lambda x: -x["n_listed"])
    # Realized monthly ROI: what your capital has actually earned per month, across real round trips.
    rois = []
    for o in rows:
        d = observed_days(o)
        if d is None or o.get("sale_price") is None or not o.get("bought_price"):
            continue
        rois.append(((o["sale_price"] - o["bought_price"]) / o["bought_price"]) * (30 / max(d, 1)))
    med_roi = _median(rois)
    return {
        "global": _calibrate_liquidity_group(GLOBAL, rows, s, now),
        "categories": categories,
        "realized_monthly_roi": round(med_roi, 3) if med_roi is not None else None,
    }


def liquidity_adjustment_for(report: dict[str, Any] | None, category: str) -> dict[str, Any]:
    """The speed adjustment for a new estimate in this category. Category first, then global."""
    none = {"days_multiplier": 1.0, "measured_n": 0}
    if not report:
        return none
    cats = report.get("categories", [])
    cat = next((x for x in cats if x["category"] == (category or "Uncategorized")), None)
    pick = cat if cat and cat["basis"] == "measured" else (
        report["global"] if report.get("global", {}).get("basis") == "measured" else None)
    if not pick:
        return none
    return {"days_multiplier": pick["days_multiplier"], "measured_n": pick["n_sold"]}


def build_report(outcomes: Iterable[dict[str, Any]], s: Settings, now: float | None = None) -> dict[str, Any]:
    now = now or time.time()
    rows = list(outcomes)
    by_cat: dict[str, list[dict[str, Any]]] = {}
    for o in rows:
        by_cat.setdefault(o.get("category") or "Uncategorized", []).append(o)
    categories = sorted((_calibrate_group(c, r, s, now) for c, r in by_cat.items()),
                        key=lambda x: -x["n_closed"])
    sold = [o for o in rows if o.get("sale_price") is not None]
    realized = sum((o["sale_price"] - o["bought_price"]) for o in sold if o.get("bought_price") is not None)
    return {
        "generated_at": now,
        "global": _calibrate_group(GLOBAL, rows, s, now),
        "categories": categories,
        "totals": {"closed": len(rows), "sold": len(sold),
                   "realized_profit": round(realized, 2) if sold else None},
        "liquidity": build_liquidity_report(rows, s, now),
    }


def adjustment_for(report: dict[str, Any] | None, category: str, s: Settings) -> dict[str, Any]:
    """The adjustment to apply to a new valuation. Falls back to global, then to a no-op."""
    none = {"bias": 1.0, "confidence_factor": 1.0, "basis": "none", "n": 0}
    if not s.calibration or not report:
        return none
    cat = next((x for x in report.get("categories", []) if x["category"] == (category or "Uncategorized")), None)
    pick = cat if cat and cat["basis"] != "none" else (report["global"] if report["global"]["basis"] != "none" else None)
    if not pick:
        return none
    return {"bias": pick["bias"], "confidence_factor": pick["confidence_factor"],
            "basis": pick["basis"], "n": pick["n_closed"]}

File: lib/spread/calibration.ts
// The feedback loop: grade past predictions against what actually happened, then bend future
// valuations toward reality.
//
// Two tiers of evidence, weakest first:
//
//   1. HAMMER PRICES (free, automatic, every closed lot — even ones you never bid on).
//      HiBid publishes `priceRealized`. A hammer price is not a resale value, so it cannot tell us
//      we were too LOW. But it can prove we were too HIGH: if a lot's landed cost at the hammer
//      meets or beats the net resale we predicted, then either the winner overpaid or — far more
//      often across many lots — our number was inflated. The share of lots where that happens is
//      the `overshoot_rate`, and it is a provably-wrong rate, not a guess.
//
//   2. REAL SALES (ground truth, entered when you actually sell something).
//      actual sale / predicted mid. This is the real calibration signal and it overrides tier 1
//      as soon as there is enough of it.
//
// Bias is clamped and requires a minimum sample so a thin or unlucky week cannot swing the model.

import type { Config } from "./config";
import type { CalibrationReport, CategoryCalibration, CategoryLiquidity, LiquidityReport, Outcome } from "./types";

const GLOBAL = "__all__";

function median(xs: number[]): number | null {
  if (!xs.length) return null;
  const s = [...xs].sort((a, b) => a - b);
  const m = Math.floor(s.length / 2);
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
}

/** Landed cost at the hammer, over the net resale we predicted. >= 1 means the deal was never there. */
export function hammerRatio(o: Outcome): number | null {
  if (!o.landed_at_hammer || !o.predicted_net || o.predicted_net <= 0) return null;
  return o.landed_at_hammer / o.predicted_net;
}

/** Actual sale price over the mid we predicted. 1.0 is a perfect call. */
export function saleRatio(o: Outcome): number | null {
  if (!o.sale_price || !o.predicted_mid || o.predicted_mid <= 0) return null;
  return o.sale_price / o.predicted_mid;
}

function calibrateGroup(category: string, rows: Outcome[], c: Config, now: number): CategoryCalibration {
  const hammers = rows.map(hammerRatio).filter((x): x is number => x !== null && Number.isFinite(x));
  const sales = rows.map(saleRatio).filter((x): x is number => x !== null && Number.isFinite(x));
  const medHammer = median(hammers);
  const medSale = median(sales);
  const overshoot = hammers.length ? hammers.filter((r) => r >= 1).length / hammers.length : null;
  const mape = sales.length ? median(sales.map((r) => Math.abs(r - 1))) : null;

  let bias = 1;
  let confidenceFactor = 1;
  let basis: CategoryCalibration["basis"] = "none";

  if (sales.length >= c.calibrationMinSales && medSale !== null) {
    // Ground truth wins outright.
    basis = "sales";
    bias = medSale;
    // Wide dispersion against real sales means the number is a coin flip; discount confidence.
    if (mape !== null) confidenceFactor = Math.max(0.4, Math.min(1, 1 - Math.max(0, mape - 0.2)));
  } else if (hammers.length >= c.calibrationMinClosed && overshoot !== null) {
    // No sales yet. We can only detect inflation, so only ever haircut — never inflate — on this basis.
    basis = "hammer";
    const excess = Math.max(0, overshoot - 0.25); // a quarter of picks getting outbid is normal
    bias = Math.min(1, 1 - excess);
    confidenceFactor = Math.max(0.5, Math.min(1, 1 - excess));
  }

  bias = Math.max(c.calibrationMinBias, Math.min(c.calibrationMaxBias, bias));
  return {
    category,
    n_closed: rows.length,
    n_sold: sales.length,
    median_hammer_ratio: medHammer,
    overshoot_rate: overshoot,
    median_sale_ratio: medSale,
    sale_mape: mape,
    bias: Math.round(bias * 1000) / 1000,
    confidence_factor: Math.round(confidenceFactor * 1000) / 1000,
    basis,
    updated_at: now,
  };
}

// ---------------------------------------------------------------------------- liquidity feedback
//
// The price loop above asks "was the number right". This one asks "did it ever actually sell", which is
// the question a $40 item sitting unviewed for three months answers differently.
//
// Two measurements, and the second is the one that keeps you honest:
//
//   OBSERVED DAYS — for the things that sold, listed_at to sale_at. Median observed over median predicted
//   gives a per-category multiplier that stretches (or compresses) every future speed estimate.
//
//   30-DAY SELL RATE — sold within 30 days, over everything that had a fair shot at 30 days. The
//   denominator deliberately includes listings that are STILL SITTING, because those are the whole point:
//   a loop that only learns from things that sold would conclude everything sells. A listing that has been
//   up for 12 days and has not sold is not yet evidence either way, so it is excluded until day 30
//   (right-censoring); one that has been up for 90 days unsold counts fully against the rate.

/** Days from listing to sale, for the ones that sold. */
export function observedDays(o: Outcome): number | null {
  if (!o.listed_at || !o.sale_at) return null;
  const d = (o.sale_at - o.listed_at) / 86400;
  return d >= 0 && Number.isFinite(d) ? d : null;
}

/** Days a listing has been up, sold or not. */
export function daysOnMarket(o: Outcome, now: number): number | null {
  if (!o.listed_at) return null;
  const end = o.sale_at ?? now;
  const d = (end - o.listed_at) / 86400;
  return d >= 0 && Number.isFinite(d) ? d : null;
}

function calibrateLiquidityGroup(category: string, rows: Outcome[], c: Config, now: number): CategoryLiquidity {
  const listed = rows.filter((o) => !!o.listed_at);
  const sold = listed.filter((o) => o.sale_at !== null && o.sale_price !== null);
  const obs = sold.map(observedDays).filter((x): x is number => x !== null);
  // Compare like with like: only the items we both predicted and observed.
  const pairs = sold
    .map((o) => ({ obs: observedDays(o), pred: o.predicted_days }))
    .filter((x): x is { obs: number; pred: number } => x.obs !== null && !!x.pred && x.pred > 0);
  const medObs = median(obs);
  const medPred = median(pairs.map((x) => x.pred));

  let multiplier = 1;
  let basis: CategoryLiquidity["basis"] = "none";
  if (pairs.length >= c.liquidityMinSales) {
    const ratios = pairs.map((x) => x.obs / x.pred).filter((r) => Number.isFinite(r) && r > 0);
    const m = median(ratios);
    if (m !== null) {
      multiplier = Math.max(0.3, Math.min(4, m));
      basis = "measured";
    }
  }

  // Censoring: a listing only enters the 30-day denominator once it has had 30 days, or once it sold.
  const eligible = listed.filter((o) => {
    const age = daysOnMarket(o, now);
    return age !== null && (o.sale_at !== null || age >= 30);
  });
  const within30 = eligible.filter((o) => {
    const d = observedDays(o);
    return d !== null && d <= 30;
  });
  const stuck = listed.filter((o) => o.sale_at === null && (daysOnMarket(o, now) ?? 0) >= 60).length;

  return {
    category,
    n_listed: listed.length,
    n_sold: sold.length,
    observed_days: medObs === null ? null : Math.round(medObs * 10) / 10,
    predicted_days: medPred === null ? null : Math.round(medPred * 10) / 10,
    days_multiplier: Math.round(multiplier * 1000) / 1000,
    sell_rate_30d: eligible.length ? Math.round((within30.length / eligible.length) * 1000) / 1000 : null,
    stuck,
    basis,
    updated_at: now,
  };
}

function buildLiquidityReport(outcomes: Outcome[], c: Config, now: number): LiquidityReport {
  const byCat = new Map<string, Outcome[]>();
  for (const o of outcomes) {
    const k = o.category || "Uncategorized";
    (byCat.get(k) ?? byCat.set(k, []).get(k)!).push(o);
  }
  const categories = [...byCat.entries()]
    .map(([cat, rows]) => calibrateLiquidityGroup(cat, rows, c, now))
    .filter((x) => x.n_listed > 0)
    .sort((a, b) => b.n_listed - a.n_listed);
  // Realized monthly ROI: what your capital has actually earned per month, across real round trips.
  const rois = outcomes
    .map((o) => {
      const d = observedDays(o);
      if (d === null || o.sale_price === null || !o.bought_price || o.bought_price <= 0) return null;
      return ((o.sale_price - o.bought_price) / o.bought_price) * (30 / Math.max(d, 1));
    })
    .filter((x): x is number => x !== null && Number.isFinite(x));
  const medRoi = median(rois);
  return {
    global: calibrateLiquidityGroup(GLOBAL, outcomes, c, now),
    categories,
    realized_monthly_roi: medRoi === null ? null : Math.round(medRoi * 1000) / 1000,
  };
}

/** The speed adjustment to apply to a new estimate in this category. Category first, then global. */
export function liquidityAdjustmentFor(report: LiquidityReport | null | undefined, category: string): { daysMultiplier: number; measuredN: number } {
  const none = { daysMultiplier: 1, measuredN: 0 };
  if (!report) return none;
  const cat = report.categories.find((x) => x.category === (category || "Uncategorized"));
  const pick = cat && cat.basis === "measured" ? cat : report.global.basis === "measured" ? report.global : null;
  return pick ? { daysMultiplier: pick.days_multiplier, measuredN: pick.n_sold } : none;
}

export function buildReport(outcomes: Outcome[], c: Config, now = Date.now() / 1000): CalibrationReport {
  const byCat = new Map<string, Outcome[]>();
  for (const o of outcomes) {
    const k = o.category || "Uncategorized";
    (byCat.get(k) ?? byCat.set(k, []).get(k)!).push(o);
  }
  const categories = [...byCat.entries()]
    .map(([cat, rows]) => calibrateGroup(cat, rows, c, now))
    .sort((a, b) => b.n_closed - a.n_closed);
  const realized = outcomes
    .filter((o) => o.sale_price !== null && o.bought_price !== null)
    .reduce((a, o) => a + (o.sale_price! - o.bought_price!), 0);
  return {
    generated_at: now,
    global: calibrateGroup(GLOBAL, outcomes, c, now),
    categories,
    totals: {
      closed: outcomes.length,
      sold: outcomes.filter((o) => o.sale_price !== null).length,
      realized_profit: outcomes.some((o) => o.sale_price !== null) ? Math.round(realized * 100) / 100 : null,
    },
    liquidity: buildLiquidityReport(outcomes, c, now),
  };
}

/** The adjustment to apply to a new valuation in this category. Falls back to global, then to no-op. */
export function adjustmentFor(report: CalibrationReport | null, category: string, c: Config): { bias: number; confidence_factor: number; basis: string; n: number } {
  const none = { bias: 1, confidence_factor: 1, basis: "none", n: 0 };
  if (!c.calibration || !report) return none;
  const cat = report.categories.find((x) => x.category === (category || "Uncategorized"));
  const pick = cat && cat.basis !== "none" ? cat : report.global.basis !== "none" ? report.global : null;
  if (!pick) return none;
  return { bias: pick.bias, confidence_factor: pick.confidence_factor, basis: pick.basis, n: pick.n_closed };
}
